syntax = "proto3";

service P2PSync{
    rpc PeerListUpdate(PeerUpdate) returns (Empty) {}
    rpc ListenCommands(Empty) returns (stream DatabaseCommand) {}
    rpc Connect(Peer) returns (Empty) {}
    rpc Heartbeat(Empty) returns (Empty) {}
    rpc SendCommand(DatabaseCommand) returns (Empty) {}
    rpc RequestPeerList(Peer) returns (PeerList) {}
}

message Empty {
}

message PeerUpdate {
    Peer peer = 1;
    bool add = 2;
}

message PeerList {
    repeated Peer peers = 1;
}

message Peer {
    string host = 1;
    string port = 2;
}

message DatabaseCommand {
    int32 timestamp = 1;
    string command = 2;
}
